source/common/conn_pool/conn_pool_base.cc

```diff
@@ -425,19 +424,13 @@ void ConnPoolImplBase::checkForIdleAndCloseIdleConnsIfDraining() {
 
 void ConnPoolImplBase::onConnectionEvent(ActiveClient& client, absl::string_view failure_reason,
                                          Network::ConnectionEvent event) {
-  if (client.state() == ActiveClient::State::CONNECTING) {
-    ASSERT(connecting_stream_capacity_ >= client.effectiveConcurrentStreamLimit());
-    connecting_stream_capacity_ -= client.effectiveConcurrentStreamLimit();
-  }
-
-  if (client.connect_timer_) {
```

Comment:
why move this? it is still done in all the cases.  are you adding a new connection event for 0-rtt?

ohhh, we did this because isConnecting relies on the timer rather than ActiveClient::State::CONNECTING?  That seems kind of fragile.  Would it make sense to instead add another active client state for 0-rtt and leave isConnecting to be based on CONNECTING?  Honestly not sure, as I haven't thought out the next two PRs you probably have in the works :-)

Reply:
Another state for 0-RTT probably won't help here because the different states have to be mutually exclusive. But a connecting(0-RTT connected) client can become BUSY or DRAINING. So the intention for this PR is to make isConnecting() not to based on the state at all. Right now, what I could think of is relying on the connect_timer_ as H3 pool actually checks it to determine if the connection is connected or not. If you think it is fragile, maybe we can add another bool to indicate that. WDYT?

Reply:
I have the state transition with 0-RTT supported clients in https://github.com/envoyproxy/envoy/pull/20167/files#diff-1ca21bd145a8c09791b9659ca8cdfbec5dfb86067b8f4615981ae618ac31cc7f.